src/source.ts

```diff
@@ -138,12 +140,24 @@ export class Source<T extends IAnyModelType, Q extends Query>
     ) {
       return result.values;
     }
-    const items = await this._load(q);
+    const items = await this.loadOrExisting(key, q).finally(() =>
+      this._existingLoad.delete(key)
+    );
     const values = items.map((item: any) => this.addOrUpdate(item));
     this.setCache(key, values, timestamp);
     return values;
   }
 
+  async loadOrExisting(key: string, q: Q): Promise<SnapshotIn<T>[]> {
```

Comment:
I'm not a big fan of this signature: key is derived from q, and here it's repeated. I realize it's slightly more performant than recalculating 'key' from 'q' in this method, but it would make it a lot cleaner that this is really a load that recycles promises.

I think the responsibility isn't really spread properly. Either you make it a full drop-in replacement of _load, or you handle all the promise caching within load(), but now you do *some* of the promise caching in here and the cleanup inside load itself.

I would suggest something like this:

```typescript
async loadReusePromise(q: Q): Promise<SnapshotIn<T>[]> {
   const key = this._keyForQuery(q);
    const promise = this._loadPromises.get(key);
    if (promise != null) {
      return promise;
   }
  const promise = this._load(q).finally(() => this._loadPromises.delete(key));
  this._loadPromises.set(key, promise);
  return promise;
}
```

Doing the finally in the place I suggested should work, as every new promise passes through there.